src/models/ili9342c.rs

```diff
@@ -0,0 +1,109 @@
+use display_interface::{DataFormat, DisplayError, WriteOnlyDataCommand};
+use embedded_graphics_core::{pixelcolor::Rgb565, prelude::IntoStorage};
+use embedded_hal::{blocking::delay::DelayUs, digital::v2::OutputPin};
+
+use crate::{instruction::Instruction, Display, DisplayOptions, Error, Orientation};
+
+use super::{write_command, Model};
+
+/// ILI9342C display with Reset pin
+/// in Rgb565 color mode
+/// Backlight pin is not controlled
+pub struct ILI9342CRgb565;
+
+impl Model for ILI9342CRgb565 {
+    type ColorFormat = Rgb565;
+
+    fn new() -> Self {
+        Self
+    }
+
+    fn init<RST, DELAY, DI>(
+        &mut self,
+        di: &mut DI,
+        rst: &mut Option<RST>,
+        delay: &mut DELAY,
+        options: DisplayOptions,
+    ) -> Result<u8, Error<RST::Error>>
+    where
+        RST: OutputPin,
+        DELAY: DelayUs<u32>,
+        DI: WriteOnlyDataCommand,
+    {
+        match rst {
+            Some(ref mut rst) => self.hard_reset(rst, delay)?,
+            None => write_command(di, Instruction::SWRESET, &[])?,
+        }
+
+        delay.delay_us(120_000);
+
+        init_common(di, delay, options).map_err(|_| Error::DisplayError)
+    }
+
+    fn write_pixels<DI, I>(&mut self, di: &mut DI, colors: I) -> Result<(), DisplayError>
+    where
+        DI: WriteOnlyDataCommand,
+        I: IntoIterator<Item = Self::ColorFormat>,
+    {
+        write_command(di, Instruction::RAMWR, &[])?;
+        let mut iter = colors.into_iter().map(|c| c.into_storage());
+
+        let buf = DataFormat::U16BEIter(&mut iter);
+        di.send_data(buf)
+    }
+
+    fn display_size(&self, orientation: Orientation) -> (u16, u16) {
+        match orientation {
+            Orientation::Portrait(_) | Orientation::PortraitInverted(_) => (320, 240),
+            Orientation::Landscape(_) | Orientation::LandscapeInverted(_) => (240, 320),
+        }
+    }
+}
+
+// simplified constructor for Display
+
+impl<DI, RST> Display<DI, RST, ILI9342CRgb565>
+where
+    DI: WriteOnlyDataCommand,
+    RST: OutputPin,
+{
+    ///
+    /// Creates a new [Display] instance with [ILI9342C] as the [Model]
+    /// *WARNING* Rgb565 only works on non-SPI setups with the ILI9342C!
+    ///
+    /// # Arguments
+    ///
+    /// * `di` - a [DisplayInterface](WriteOnlyDataCommand) for talking with the display
+    /// * `rst` - display hard reset [OutputPin]
+    /// * `model` - the display [Model]
+    ///
+    pub fn ili9342c_rgb565(di: DI, rst: RST) -> Self {
+        Self::with_model(di, Some(rst), ILI9342CRgb565::new())
+    }
+}
+
+// common init for all color format models
+fn init_common<DELAY, DI>(
+    di: &mut DI,
+    delay: &mut DELAY,
+    options: DisplayOptions,
+) -> Result<u8, DisplayError>
+where
+    DELAY: DelayUs<u32>,
+    DI: WriteOnlyDataCommand,
+{
+    let madctl = options.madctl() ^ 0b0000_1000; // this model has flipped RGB/BGR bit;
```

Comment:
This was a mistake on my end with ST7789. Are you sure it's required for this model?

Reply:
You're right, this display is in RGB mode by default. For whatever reason my colours still seem to be displaying using BGR (setting the background to red displays as blue) without that, so I think this is why I had set that initially. I will dig into the datasheet a bit more and try to figure out what's going on.

Reply:
I'm not really sure what's going on here, as far as I can tell the display is configured correctly. I am unable to get it to work properly without inverting this bit, the byte order is always backwards. I did find a couple more drivers for this display, and they both seem to be setting the RGB/BGR bit to 1 (BGR), so I'm not totally crazy at least (well, I hope).

- https://github.com/retro-esp32/Odroid-GO-ILI9342/blob/master/ILI9342/Arduino%20Libraries/Adafruit_ILI9342/Adafruit_ILI9342.cpp#L160  
- https://github.com/MilovdZee/TFTLCD-Library_ILI9342/blob/master/Adafruit_TFTLCD_ILI9342.cpp#L150

Not sure how you'd like to proceed on this. I can update the comment if that is adequate, otherwise I'm not sure what to do.

Reply:
This is interesting, the first driver inverts RGB but also sets MX bit. The second driver also inverts RGB but then sets the MY bit. So they don't even agree between each other. I guess one is aiming for portrait mirror image, while the other goes for inverted portrait orientation.